### PR TITLE
isotpdump: option to capture functional addressing traffic

### PR DESCRIPTION
**Pull request by uuie**
add option '-b' to capture functional broadcast data eg: `-b 98DAFFF1` or `-b 18DAFFF1`

there are 3 can_id used during FOTA by UDS over J1939: 
 `18DATTSS`, `18DASSTT` and `18DBFFSS` (or `18DB33SS`, functional addressing)
value `TT` is the target address and `SS` is the source address for J1939 can_id. 

add this option to capture all these traffic

### CODE REVIEW

File: isotpdump.c [hunk 1]
```diff
@@ -78,6 +78,7 @@ void print_usage(char *prg)
 	fprintf(stderr, "         -a           (print data also in ASCII-chars)\n");
 	fprintf(stderr, "         -t <type>    (timestamp: (a)bsolute/(d)elta/(z)ero/(A)bsolute w date)\n");
 	fprintf(stderr, "         -u           (print uds messages)\n");
+	fprintf(stderr, "         -b [can_id]  (broadcast can_id, eg: 18DBFFF1)\n");
```
Comment (marckleinebudde) [inline isotpdump.c:81]: IMHO it makes more sense to move this right after the `-d` parameter.

File: isotpdump.c [hunk 2]
```diff
@@ -273,12 +275,16 @@ int main(int argc, char **argv)
 		case 'u':
 		        uds_output = 1;
 			break;
+		case 'b':
```
Comment (marckleinebudde) [inline isotpdump.c:278]: Move this after 'd'.

File: isotpdump.c [hunk 3]
```diff
@@ -273,12 +275,16 @@ int main(int argc, char **argv)
...
+			bst = strtoul(optarg, NULL, 16);
+			if (strlen(optarg) > 7)
+				bst |= CAN_EFF_FLAG;
+			break;
 
 		case '?':
 			print_usage(basename(argv[0]));
 			exit(0);
 			break;
-
```
Comment (marckleinebudde) [inline isotpdump.c:281]: nitpick: unrelated change, please remove

File: isotpdump.c [hunk 4]
```diff
@@ -320,8 +326,20 @@ int main(int argc, char **argv)
...
+		rfilter[2].can_id   = bst & (CAN_EFF_MASK | CAN_EFF_FLAG);
+		rfilter[2].can_mask = (CAN_EFF_MASK|CAN_EFF_FLAG|CAN_RTR_FLAG);
+	} else {
+		rfilter[2].can_id   = bst & CAN_SFF_MASK;
+		rfilter[2].can_mask = (CAN_SFF_MASK|CAN_EFF_FLAG|CAN_RTR_FLAG);
+	}
+
+	if (bst != NO_CAN_ID) {
+		setsockopt(s, SOL_CAN_RAW, CAN_RAW_FILTER, &rfilter, sizeof(rfilter));
+	}else {
```
Comment (marckleinebudde) [inline isotpdump.c:339]:
```suggestion
	} else {
```

**Review by marckleinebudde — CHANGES_REQUESTED** [review]

**Comment by hartkopp** [comment]
Hi,
thanks for your contribution.

Besides some small editorial issues (e.g. missing color support and different help text pattern) I have a question about your "3 can_id used during FOTA by UDS over J1939" ...

> add option '-b' to capture functional broadcast data eg: `-b 98DAFFF1` or `-b 18DAFFF1`
> 
> there are 3 can_id used during FOTA by UDS over J1939: `18DATTSS`, `18DASSTT` and `18DBFFSS` (or `18DB33SS`, functional addressing) add this option to capture all these traffic

If you want to see these CAN IDs (what dies "SS" and "TT" mean?) won't it make more sense to have a mask/value filter like we have in `candump` ?

E.g. `-b 18DA0000:9FFF0000`

**Comment by uuie** [comment]
@hartkopp  thanks for your advice :)
these ascii and uds message privided by `isotpdump` help me a lot during uds debugging, which i'm not found in `candump`. 
image

sorry for the `SS` `TT` description, i'll change them to `SRC_ADDR` `DST_ADDR`.

File: isotpdump.c [hunk 5]
```diff
@@ -320,8 +328,20 @@ int main(int argc, char **argv)
 		rfilter[1].can_id   = dst & CAN_SFF_MASK;
 		rfilter[1].can_mask = (CAN_SFF_MASK|CAN_EFF_FLAG|CAN_RTR_FLAG);
 	}
+	if (bst & CAN_EFF_FLAG) {
```
Comment (marckleinebudde) [inline isotpdump.c:331]:
```suggestion

	if (bst & CAN_EFF_FLAG) {
```

File: isotpdump.c [hunk 6]
```diff
@@ -320,8 +328,20 @@ int main(int argc, char **argv)
...
+	} else {
+		rfilter[2].can_id   = bst & CAN_SFF_MASK;
+		rfilter[2].can_mask = (CAN_SFF_MASK|CAN_EFF_FLAG|CAN_RTR_FLAG);
+	}
+
+	if (bst != NO_CAN_ID) {
+		setsockopt(s, SOL_CAN_RAW, CAN_RAW_FILTER, &rfilter, sizeof(rfilter));
+	} else {
+		setsockopt(s, SOL_CAN_RAW, CAN_RAW_FILTER, &rfilter, sizeof(rfilter) - sizeof(rfilter[0]));
+	}
```
Comment (marckleinebudde) [inline isotpdump.c:343]:
```suggestion
	if (bst != NO_CAN_ID)
		setsockopt(s, SOL_CAN_RAW, CAN_RAW_FILTER, &rfilter, sizeof(rfilter));
	else
		setsockopt(s, SOL_CAN_RAW, CAN_RAW_FILTER, &rfilter, sizeof(rfilter) - sizeof(rfilter[0]));
```

**Review by marckleinebudde — CHANGES_REQUESTED** [review]

File: isotpdump.c (hunk 2 above)
Comment (uuie) [reply isotpdump.c]: fixed :)

**Comment by hartkopp** [comment]
Thanks for the update. The patch looks better now.

> sorry for the `SS` `TT` description, i'll change them to `SRC_ADDR` `DST_ADDR`.

???

I can understand, why you would  like to have an additional broadcast CAN ID.

But my question remains valid, if it would make more sense to have a filter definition for the broadcast CAN IDs instead of having just a single CAN ID. What do you think about this? Did you get what I am suggesting?

File: isotpdump.c (hunk 5 above)
Comment (uuie) [reply isotpdump.c:332]: additional nl added :)

File: isotpdump.c (hunk 6 above)
Comment (uuie) [reply isotpdump.c]: fixed :)

**Comment by uuie** [comment]
> I can understand, why you would like to have an additional broadcast CAN ID.

i'm perform uds fota over J1939. 
some key steps including:
SS=tester addr
TT=ecu addr
1. switch ecu to extended session ($10 03) using functional addressing from tester,  can_id=18DBFFSSx
2. switch ecu to programing session (10 02) using phy addressing from tester,  can_id=18DADDSSx
3.  ecu answer each uds request to tester using phy addressing, can_id=18DASSDDx

i could not capture both functional and phy traffic send by tester before.

**Comment by hartkopp** [comment]
Ok, so a _**single**_ CAN ID to display the functional request is sufficient for your use-case, right?

**Review by marckleinebudde — APPROVED** [review]
